Expand build argument from environment when no value specified

antechrestos opened the pull request:
Fixes #713 

**Description**

This change adds the feature of getting value of build argument from environment when build argument is specified as follows

`/kaniko/executor .... --build-arg MY_ARGUMENT ...` 



**Submitter Checklist**


- [x] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [x] Adds integration tests if needed.

**Reviewer Notes**

- [x] The code flow looks good. 
- [x] Unit tests and or integration tests added.


**Release Notes**

```
Allow user to provide build argument from environment variable as follows:
$> export MY_ARGUMENT="the value"
$> /kaniko/executor ... --build-arg MY_ARGUMENT ...
Or
$> docker run -e MY_ARGUMENT="the value" gcr.io/kaniko-project/executor ... --build-arg MY_ARGUMENT ...
```

antechrestos commented:
@cvgw As promised... Do I need to add integration test?

tejal29 commented:
Thank you so much for your contribution!

stewi2 commented:
> won't `docker -e SOME_SECRET=secret-value run ....` output the value of the secret to the log/shell history?

At least that part can be solved via a .env file. I.e.

```
docker --env-file .secrets run ....
```

with `.secrets` containing this:

```
SOME_SECRET=secret-value
```

Granted, that file would need to be written to disk first, but this is a common way to pass along environment variables without exposing them via the command line.

antechrestos commented:
@tejal29 you're welcome
@cvgw I added it. To do so I added a new dedicated dockerfile. I also added a way to pass environment variable to build (with `cmd.Env` for docker build and `-e` for kaniko build as kaniko build is done inside docker) and also tested the expansion of docker variable.
For this test a check of output is provided te do a research on argument value and ensure that it is not printed.

cvgw commented on cmd/executor/cmd/root.go:
```diff
@@ -206,6 +207,17 @@ func resolveDockerfilePath() error {
 }
 
+// resolveEnvironmentBuildArgs replace build args without value by the same named environment variable
+func resolveEnvironmentBuildArgs(arguments []string, resolver func(string) string) {
```
So this will take an argument like `FOO` and replae it with `FOO=FOO` ?

antechrestos replied:
@cvgw this will take the argument array such as `[ "foo=bar", "EnvVariable" ]` crawl through it and replace any argument without `=` with `"EnvVariable=EnvValue"` **if** a  value was returned by `resolver` function  (`os.Getenv` in prod code, mocked function in tests). If zero length value returned, it will be replaced with `"EnvVariable="`

cvgw replied:
Thanks for explaining, cheers

cvgw commented on integration/images.go:
```diff
@@ -71,6 +78,36 @@ var additionalKanikoFlagsMap = map[string][]string{
+
+// Checks if argument are not printed in output.
+// Argument may be passed through --build-arg key=value manner or --build-arg key with key in environment
+func checkArgsNotPrinted(dockerfile string, out []byte) error {
```
So this checks to make sure the value of none of the build args are logged? It makes no distinction between whether they were added as literals or expanded from the env?

antechrestos replied:
@cvgw I could have make the distinction in the test, however I could not find any way to make the distinction in the code as env like variable (env, arg and meta arg ) are used everywhere as a mere string `key=value` and could not make the distinction without breaking everything.
Do you want I change the test? Or everything?

cvgw replied:
No changes needed. Was just clarifying for my owner understanding. I think it's reasonable to not log any ARG values.

antechrestos commented:
@cvgw something went wrong with kaniko. However, things went well in Travis 🤔...

tejal29 commented:
@antechrestos i triggered a rebuilt. Previous failure was a flake.

antechrestos commented:
@tejal29 thanks

antechrestos commented:
@tejal29 😢

antechrestos commented:
@tejal29  I have rebased my branch. I am not sure it might change anything :thinking:

antechrestos commented:
@cvgw @tejal29 may I have another try with kokoro? :pray:

tejal29 commented:
@antechrestos The travis test passed, so i am going to merge this in!

antechrestos commented:
@tejal29 at our own risk 😊

sourcec0de commented:
THANK YOU! ❤️

nhed commented:
Thanks this is so helpful (albeit undocumented) feature
